File: final kel/lost-and-found/python-packages/kelly 2/src/kelly/engines/rhythm_engine.py
"""Rhythm Engine - Generates emotion-driven drum patterns.

Creates drum patterns that express emotional intent through
density, accent patterns, and ghost notes.
"""

from dataclasses import dataclass, field
from typing import List, Dict, Optional, Tuple
from enum import Enum
import random

TICKS_PER_BEAT = 480

# GM Drum Map
GM_KICK = 36
GM_SNARE = 38
GM_CLOSED_HH = 42
GM_OPEN_HH = 46
GM_CRASH = 49
GM_RIDE = 51
GM_TOM_LOW = 45
GM_TOM_MID = 47
GM_TOM_HIGH = 50


class DrumPattern(Enum):
    """Drum pattern archetypes."""
    FOUR_ON_FLOOR = "four_on_floor"
    BACKBEAT = "backbeat"
    HALFTIME = "halftime"
    BREAKBEAT = "breakbeat"
    LINEAR = "linear"
    SPARSE = "sparse"
    TRIBAL = "tribal"
    SHUFFLE = "shuffle"


@dataclass
class RhythmConfig:
    """Configuration for rhythm generation."""
    emotion: str = "neutral"
    bars: int = 4
    tempo_bpm: int = 120
    time_signature: Tuple[int, int] = (4, 4)
    pattern_override: Optional[DrumPattern] = None
    include_hihats: bool = True
    include_fills: bool = True
    seed: int = -1


@dataclass
class DrumHit:
    """A single drum hit."""
    pitch: int
    start_tick: int
    duration_ticks: int
    velocity: int
    is_ghost: bool = False


@dataclass
class RhythmOutput:
    """Complete rhythm generation output."""
    hits: List[DrumHit]
    pattern: DrumPattern
    emotion: str
    bars: int


EMOTION_PROFILES = {
    "grief": {
        "pattern": DrumPattern.SPARSE,
        "kick_density": 0.3,
        "snare_density": 0.2,
        "hihat_density": 0.4,
        "velocity_range": (35, 60),
        "ghost_probability": 0.05,
        "fill_probability": 0.1,
    },
    "sadness": {
        "pattern": DrumPattern.HALFTIME,
        "kick_density": 0.4,
        "snare_density": 0.25,
        "hihat_density": 0.5,
        "velocity_range": (40, 70),
        "ghost_probability": 0.1,
        "fill_probability": 0.15,
    },
    "anger": {
        "pattern": DrumPattern.FOUR_ON_FLOOR,
        "kick_density": 0.9,
        "snare_density": 0.7,
        "hihat_density": 1.0,
        "velocity_range": (90, 127),
        "ghost_probability": 0.25,
        "fill_probability": 0.3,
    },
    "anxiety": {
        "pattern": DrumPattern.BREAKBEAT,
        "kick_density": 0.6,
        "snare_density": 0.5,
        "hihat_density": 0.8,
        "velocity_range": (60, 95),
        "ghost_probability": 0.2,
        "fill_probability": 0.25,
    },
    "joy": {
        "pattern": DrumPattern.BACKBEAT,
        "kick_density": 0.7,
        "snare_density": 0.5,
        "hihat_density": 0.9,
        "velocity_range": (70, 100),
        "ghost_probability": 0.15,
        "fill_probability": 0.2,
    },
    "hope": {
        "pattern": DrumPattern.BACKBEAT,
        "kick_density": 0.6,
        "snare_density": 0.4,
        "hihat_density": 0.7,
        "velocity_range": (55, 85),
        "ghost_probability": 0.1,
        "fill_probability": 0.15,
    },
    "defiance": {
        "pattern": DrumPattern.BREAKBEAT,
        "kick_density": 0.8,
        "snare_density": 0.6,
        "hihat_density": 0.85,
        "velocity_range": (80, 115),
        "ghost_probability": 0.2,
        "fill_probability": 0.25,
    },
    "emptiness": {
        "pattern": DrumPattern.SPARSE,
        "kick_density": 0.2,
        "snare_density": 0.1,
        "hihat_density": 0.3,
        "velocity_range": (25, 50),
        "ghost_probability": 0.02,
        "fill_probability": 0.05,
    },
    "nostalgia": {
        "pattern": DrumPattern.SHUFFLE,
        "kick_density": 0.5,
        "snare_density": 0.4,
        "hihat_density": 0.6,
        "velocity_range": (50, 80),
        "ghost_probability": 0.12,
        "fill_probability": 0.15,
    },
}


class RhythmEngine:
    """Generates emotion-driven drum patterns."""
    
    def __init__(self, seed: Optional[int] = None):
        self.profiles = EMOTION_PROFILES
        if seed is not None:
            random.seed(seed)
    
    def generate(self, config: RhythmConfig) -> RhythmOutput:
        """Generate drum pattern from config."""
        if config.seed >= 0:
            random.seed(config.seed)
        
        emotion = config.emotion.lower()
        profile = self.profiles.get(emotion, self.profiles["hope"])
        pattern = config.pattern_override or profile["pattern"]
        
        hits = []
        ticks_per_bar = TICKS_PER_BEAT * config.time_signature[0]
        
        for bar in range(config.bars):
            bar_start = bar * ticks_per_bar
            is_fill_bar = config.include_fills and bar == config.bars - 1
            
            if is_fill_bar and random.random() < profile["fill_probability"]:
                hits.extend(self._generate_fill(bar_start, ticks_per_bar, profile))
            else:
                hits.extend(self._generate_bar(
                    bar_start, ticks_per_bar, pattern, profile, config
                ))
        
        return RhythmOutput(
            hits=hits,
            pattern=pattern,
            emotion=emotion,
            bars=config.bars,
        )
    
    def _generate_bar(
        self,
        start: int,
        bar_ticks: int,
        pattern: DrumPattern,
        profile: Dict,
        config: RhythmConfig
    ) -> List[DrumHit]:
        """Generate one bar of drums."""
        hits = []
        vel_min, vel_max = profile["velocity_range"]
        sixteenth = TICKS_PER_BEAT // 4
        
        # Kick pattern
        kick_positions = self._get_kick_positions(pattern, bar_ticks)
        for pos in kick_positions:
            if random.random() < profile["kick_density"]:
                hits.append(DrumHit(
                    pitch=GM_KICK,
                    start_tick=start + pos,
                    duration_ticks=sixteenth,
                    velocity=random.randint(vel_min + 10, vel_max),
                ))
        
        # Snare pattern
        snare_positions = self._get_snare_positions(pattern, bar_ticks)
        for pos in snare_positions:
            if random.random() < profile["snare_density"]:
                hits.append(DrumHit(
                    pitch=GM_SNARE,
                    start_tick=start + pos,
                    duration_ticks=sixteenth,
                    velocity=random.randint(vel_min, vel_max),
                ))
                # Ghost note
                if random.random() < profile["ghost_probability"]:
                    ghost_pos = pos - sixteenth if pos > 0 else pos + sixteenth
                    hits.append(DrumHit(
                        pitch=GM_SNARE,
                        start_tick=start + ghost_pos,
                        duration_ticks=sixteenth // 2,
                        velocity=random.randint(vel_min // 2, vel_min),
                        is_ghost=True,
                    ))
        
        # Hi-hats
        if config.include_hihats:
            hihat_positions = self._get_hihat_positions(pattern, bar_ticks)
            for i, pos in enumerate(hihat_positions):
                if random.random() < profile["hihat_density"]:
                    is_open = pattern == DrumPattern.SHUFFLE and i % 3 == 2
                    hits.append(DrumHit(
                        pitch=GM_OPEN_HH if is_open else GM_CLOSED_HH,
                        start_tick=start + pos,
                        duration_ticks=sixteenth // 2,
                        velocity=random.randint(vel_min - 10, vel_max - 20),
                    ))
        
        return hits
    
    def _get_kick_positions(self, pattern: DrumPattern, bar_ticks: int) -> List[int]:
        """Get kick drum positions for pattern."""
        beat = TICKS_PER_BEAT
        
        if pattern == DrumPattern.FOUR_ON_FLOOR:
            return [0, beat, beat * 2, beat * 3]
        elif pattern == DrumPattern.BACKBEAT:
            return [0, beat * 2]
        elif pattern == DrumPattern.HALFTIME:
            return [0]
        elif pattern == DrumPattern.BREAKBEAT:
            return [0, int(beat * 1.5), beat * 2, int(beat * 3.5)]
        elif pattern == DrumPattern.SPARSE:
            return [0]
        elif pattern == DrumPattern.SHUFFLE:
            return [0, beat * 2]
        else:
            return [0, beat * 2]
    
    def _get_snare_positions(self, pattern: DrumPattern, bar_ticks: int) -> List[int]:
        """Get snare drum positions for pattern."""
        beat = TICKS_PER_BEAT
        
        if pattern == DrumPattern.FOUR_ON_FLOOR:
            return [beat, beat * 3]
        elif pattern == DrumPattern.BACKBEAT:
            return [beat, beat * 3]
        elif pattern == DrumPattern.HALFTIME:
            return [beat * 2]
        elif pattern == DrumPattern.BREAKBEAT:
            return [beat, int(beat * 2.75)]
        elif pattern == DrumPattern.SPARSE:
            return [beat * 2]
        elif pattern == DrumPattern.SHUFFLE:
            return [beat, beat * 3]
        else:
            return [beat, beat * 3]
    
    def _get_hihat_positions(self, pattern: DrumPattern, bar_ticks: int) -> List[int]:
        """Get hi-hat positions for pattern."""
        eighth = TICKS_PER_BEAT // 2
        sixteenth = TICKS_PER_BEAT // 4
        
        if pattern in [DrumPattern.FOUR_ON_FLOOR, DrumPattern.BACKBEAT]:
            return [i * eighth for i in range(8)]
        elif pattern == DrumPattern.HALFTIME:
            return [i * eighth for i in range(8)]
        elif pattern == DrumPattern.BREAKBEAT:
            return [i * sixteenth for i in range(16)]
        elif pattern == DrumPattern.SPARSE:
            return [0, TICKS_PER_BEAT * 2]
        elif pattern == DrumPattern.SHUFFLE:
            triplet = TICKS_PER_BEAT // 3
            return [i * triplet for i in range(12)]
        else:
            return [i * eighth for i in range(8)]
    
    def _generate_fill(
        self,
        start: int,
        bar_ticks: int,
        profile: Dict
    ) -> List[DrumHit]:
        """Generate a drum fill."""
        hits = []
        vel_min, vel_max = profile["velocity_range"]
        sixteenth = TICKS_PER_BEAT // 4
        
        # Fill in last half of bar
        fill_start = start + bar_ticks // 2
        
        toms = [GM_TOM_HIGH, GM_TOM_MID, GM_TOM_LOW]
        for i in range(8):
            pos = fill_start + i * sixteenth
            pitch = random.choice(toms + [GM_SNARE])
            hits.append(DrumHit(
                pitch=pitch,
                start_tick=pos,
                duration_ticks=sixteenth,
                velocity=random.randint(vel_min, vel_max),
            ))
        
        # Crash at end
        hits.append(DrumHit(
            pitch=GM_CRASH,
            start_tick=start + bar_ticks,
            duration_ticks=TICKS_PER_BEAT,
            velocity=vel_max,
        ))
        
        return hits


def generate_rhythm(
    emotion: str,
    bars: int = 4,
    tempo: int = 120
) -> RhythmOutput:
    """Quick rhythm generation helper."""
    engine = RhythmEngine()
    config = RhythmConfig(emotion=emotion, bars=bars, tempo_bpm=tempo)
    return engine.generate(config)

File: recovered_icloud/final_kel/src/engines/RhythmEngine.cpp
#include "RhythmEngine.h"
#include <algorithm>
#include <cmath>
#include <random>

namespace kelly {

namespace {
// Use global TICKS_PER_BEAT from KellyTypes.h instead of local definition

// Removed static RNG - each generation creates its own from seed
} // namespace

RhythmEngine::RhythmEngine() {
  initializeProfiles();
  initializeGenreModifiers();
}

void RhythmEngine::initializeProfiles() {
  // Grief - minimal, hesitant
  profiles_["grief"] = {
      {GrooveType::Human, GrooveType::Halftime, GrooveType::Broken},
      PatternDensity::Sparse,
      AccentPattern::None,
      {1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0}, // kick
      {0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0}, // snare
      0.3f,
      0.15f,
      0.1f,
      0.0f,
      {35, 70},
      0.08f,
      0.2f};

  // Sadness - slow, sparse
  profiles_["sadness"] = {
      {GrooveType::Human, GrooveType::Straight, GrooveType::Halftime},
      PatternDensity::Sparse,
      AccentPattern::Backbeat,
      {1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0},
      {0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0},
      0.4f,
      0.2f,
      0.15f,
      0.0f,
      {40, 75},
      0.06f,
      0.15f};

  // Hope - building, organic
  profiles_["hope"] = {
      {GrooveType::Straight, GrooveType::Human, GrooveType::Shuffle},
      PatternDensity::Moderate,
      AccentPattern::Backbeat,
      {1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0},
      {0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0},
      0.6f,
      0.25f,
      0.2f,
      0.1f,
      {55, 90},
      0.04f,
      0.1f};

  // Joy - energetic, bouncy
  profiles_["joy"] = {
      {GrooveType::Straight, GrooveType::Shuffle, GrooveType::Swing},
      PatternDensity::Moderate,
      AccentPattern::AllBeats,
      {1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0},
      {0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0},
      0.8f,
      0.3f,
      0.25f,
      0.15f,
      {70, 105},
      0.03f,
      0.05f};

  // Anger/Rage - aggressive, driving
  profiles_["anger"] = {
      {GrooveType::Straight, GrooveType::DoubleTime, GrooveType::FourOnFloor},
      PatternDensity::Busy,
      AccentPattern::AllBeats,
      {1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0},
      {0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1},
      0.9f,
      0.15f,
      0.3f,
      0.0f,
      {90, 127},
      0.02f,
      0.02f};
  profiles_["rage"] = profiles_["anger"];

  // Fear - erratic, tense
  profiles_["fear"] = {
      {GrooveType::Broken, GrooveType::Human, GrooveType::Trap},
      PatternDensity::Moderate,
      AccentPattern::Offbeat,
      {1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0},
      {0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0},
      0.5f,
      0.35f,
      0.15f,
      0.0f,
      {45, 85},
      0.07f,
      0.25f};

  // Anxiety - restless, unstable
  profiles_["anxiety"] = {
      {GrooveType::Broken, GrooveType::Trap, GrooveType::Human},
      PatternDensity::Busy,
      AccentPattern::Offbeat,
      {1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0},
      {0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0},
      0.75f,
      0.4f,
      0.2f,
      0.0f,
      {50, 90},
      0.05f,
      0.15f};

  // Peace - gentle, spacious
  profiles_["peace"] = {
      {GrooveType::Human, GrooveType::Swing, GrooveType::Shuffle},
      PatternDensity::Minimal,
      AccentPattern::None,
      {1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
      {0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0},
      0.2f,
      0.1f,
      0.05f,
      0.2f,
      {30, 60},
      0.06f,
      0.3f};

  // Neutral
  profiles_["neutral"] = {{GrooveType::Straight, GrooveType::Human},
                          PatternDensity::Moderate,
                          AccentPattern::Backbeat,
                          {1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0},
                          {0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0},
                          0.7f,
                          0.2f,
                          0.2f,
                          0.0f,
                          {60, 95},
                          0.02f,
                          0.05f};
}

void RhythmEngine::initializeGenreModifiers() {
  genreModifiers_["lo-fi"] = {{"velocity_mult", 0.7f},
                              {"timing_humanize_mult", 2.0f},
                              {"hat_probability_mult", 0.6f}};
  genreModifiers_["rock"] = {{"velocity_mult", 1.1f},
                             {"timing_humanize_mult", 0.8f},
                             {"hat_probability_mult", 1.0f}};
  genreModifiers_["electronic"] = {{"velocity_mult", 1.0f},
                                   {"timing_humanize_mult", 0.3f},
                                   {"hat_probability_mult", 1.2f}};
  genreModifiers_["jazz"] = {{"velocity_mult", 0.85f},
                             {"timing_humanize_mult", 1.5f},
                             {"hat_probability_mult", 0.8f},
                             {"swing_override", 0.3f}};
  genreModifiers_["hip-hop"] = {{"velocity_mult", 1.0f},
                                {"timing_humanize_mult", 1.0f},
                                {"hat_probability_mult", 1.1f}};
  genreModifiers_["ambient"] = {{"velocity_mult", 0.5f},
                                {"timing_humanize_mult", 2.5f},
                                {"hat_probability_mult", 0.3f}};
}

int RhythmEngine::applySwing(int tick, int beatTicks, float swingAmount) {
  if (swingAmount <= 0.0f)
    return tick;

  int posInBeat = tick % beatTicks;
  int eighth = beatTicks / 2;

  if (posInBeat >= eighth && posInBeat < beatTicks) {
    int offset = static_cast<int>(eighth * swingAmount);
    return tick + offset;
  }
  return tick;
}

std::vector<std::tuple<int, std::string, int>> RhythmEngine::generateHatPattern(
    int barTicks, const RhythmEmotionProfile &profile, GrooveType groove) {
  std::vector<std::tuple<int, std::string, int>> hits;
  // Create unique RNG for this function
  unsigned int seed = static_cast<unsigned int>(barTicks) ^
                      static_cast<unsigned int>(groove) ^ 0xDDDD0000;
  std::mt19937 rng(seed);
  std::uniform_real_distribution<float> dist(0.0f, 1.0f);
  std::uniform_int_distribution<int> velDist(profile.velocityRange.first - 20,
                                             profile.velocityRange.second - 10);

  int eighth = TICKS_PER_BEAT / 2;
  int numEighths = barTicks / eighth;

  for (int i = 0; i < numEighths; ++i) {
    if (dist(rng) < profile.hatProbability) {
      int tick = i * eighth;
      std::string hat = (i % 4 == 3) ? "open_hat" : "closed_hat";
      int velocity = velDist(rng);

      // Accent on beats
      if (i % 2 == 0)
        velocity += 10;
      velocity = std::clamp(velocity, 1, 127);

      hits.push_back({tick, hat, velocity});
    }
  }

  return hits;
}

std::vector<std::tuple<int, std::string, int>>
RhythmEngine::generateTrapHats(int barTicks,
                               const RhythmEmotionProfile &profile) {
  (void)profile; // Mark as intentionally unused
  std::vector<std::tuple<int, std::string, int>> hits;
  // Create unique RNG for this function
  unsigned int seed = static_cast<unsigned int>(barTicks) ^ 0xBBBB0000;
  std::mt19937 rng(seed);
  std::uniform_real_distribution<float> dist(0.0f, 1.0f);
  std::uniform_int_distribution<int> velDist(40, 100);

  int sixteenth = TICKS_PER_BEAT / 4;
  int thirtysecond = sixteenth / 2;

  // Base 16th note pattern
  for (int i = 0; i < barTicks / sixteenth; ++i) {
    if (dist(rng) < 0.8f) {
      hits.push_back({i * sixteenth, "closed_hat", velDist(rng)});
    }

    // Trap rolls on certain beats
    if ((i == 6 || i == 14) && dist(rng) < 0.6f) {
      for (int j = 0; j < 4; ++j) {
        int rollTick = i * sixteenth + j * thirtysecond;
        int vel = 60 + j * 15;
        hits.push_back({rollTick, "closed_hat", vel});
      }
    }
  }

  return hits;
}

std::vector<std::tuple<int, std::string, int>>
RhythmEngine::generateFill(int startTick, int durationTicks) {
  std::vector<std::tuple<int, std::string, int>> hits;
  // Create unique RNG for this function
  unsigned int seed = static_cast<unsigned int>(startTick) ^
                      static_cast<unsigned int>(durationTicks) ^ 0xEEEE0000;
  std::mt19937 rng(seed);

  int sixteenth = TICKS_PER_BEAT / 4;
  int numNotes = durationTicks / sixteenth;

  std::vector<std::string> fillInstruments = {"snare", "high_tom", "mid_tom",
                                              "low_tom"};

  for (int i = 0; i < numNotes; ++i) {
    int tick = startTick + i * sixteenth;
    std::string instr = fillInstruments[rng() % fillInstruments.size()];
    int velocity = 70 + static_cast<int>(40.0f * i / numNotes);
    hits.push_back({tick, instr, std::min(127, velocity)});
  }

  // Crash at end
  hits.push_back({startTick + durationTicks, "crash", 100});

  return hits;
}

RhythmOutput RhythmEngine::generate(const std::string &emotion, int bars,
                                    int tempoBpm, const std::string &genre) {
  RhythmConfig config;
  config.emotion = emotion;
  config.bars = bars;
  config.tempoBpm = tempoBpm;
  config.genre = genre;
  return generate(config);
}

RhythmOutput RhythmEngine::generate(const RhythmConfig &config) {
  // Create RNG from seed - each generation gets its own RNG
  unsigned int seed = config.seed >= 0
                          ? static_cast<unsigned int>(config.seed)
                          : static_cast<unsigned int>(std::random_device{}());
  std::mt19937 rng(seed);
  std::uniform_real_distribution<float> dist(0.0f, 1.0f);

  std::string emotionLower = config.emotion;
  std::transform(emotionLower.begin(), emotionLower.end(), emotionLower.begin(),
                 ::tolower);

  auto it = profiles_.find(emotionLower);
  auto profile = (it != profiles_.end()) ? it->second : profiles_["neutral"];

  // Apply genre modifiers
  float velocityMult = 1.0f;
  float timingMult = 1.0f;
  float hatMult = 1.0f;

  if (!config.genre.empty()) {
    auto genreIt = genreModifiers_.find(config.genre);
    if (genreIt != genreModifiers_.end()) {
      auto &mods = genreIt->second;
      if (mods.count("velocity_mult"))
        velocityMult = mods.at("velocity_mult");
      if (mods.count("timing_humanize_mult"))
        timingMult = mods.at("timing_humanize_mult");
      if (mods.count("hat_probability_mult"))
        hatMult = mods.at("hat_probability_mult");
      if (mods.count("swing_override"))
        profile.swingAmount = mods.at("swing_override");
    }
  }

  profile.hatProbability *= hatMult;
  profile.timingHumanize *= timingMult;

  GrooveType groove = config.grooveOverride.value_or(
      profile.grooves[rng() % profile.grooves.size()]);
  PatternDensity density = config.densityOverride.value_or(profile.density);

  int beatsPerBar = config.timeSignature.first;
  int ticksPerBar = beatsPerBar * TICKS_PER_BEAT;
  int totalTicks = config.bars * ticksPerBar;
  int sixteenth = TICKS_PER_BEAT / 4;

  std::vector<DrumHit> allHits;
  std::set<std::string> instrumentsUsed;

  std::uniform_int_distribution<int> velDist(
      static_cast<int>(profile.velocityRange.first * velocityMult),
      static_cast<int>(profile.velocityRange.second * velocityMult));

  for (int bar = 0; bar < config.bars; ++bar) {
    int barStart = bar * ticksPerBar;

    // Check for fill
    bool isFillBar =
        config.includeFills &&
        (std::find(config.fillOnBar.begin(), config.fillOnBar.end(), bar) !=
             config.fillOnBar.end() ||
         (bar == config.bars - 1 && dist(rng) < profile.fillProbability));

    if (isFillBar) {
      auto fillHits =
          generateFill(barStart + ticksPerBar - TICKS_PER_BEAT, TICKS_PER_BEAT);
      for (auto &[tick, instr, vel] : fillHits) {
        int midiNote = GMDrum::SNARE;
        if (instr == "high_tom")
          midiNote = GMDrum::HIGH_TOM;
        else if (instr == "mid_tom")
          midiNote = GMDrum::MID_TOM;
        else if (instr == "low_tom")
          midiNote = GMDrum::LOW_TOM;
        else if (instr == "crash")
          midiNote = GMDrum::CRASH;

        allHits.push_back({instr, midiNote, tick, sixteenth, vel, false, 0});
        instrumentsUsed.insert(instr);
      }
    }

    // Kick pattern
    for (size_t i = 0; i < profile.kickPattern.size(); ++i) {
      if (profile.kickPattern[i] && dist(rng) > profile.dropProbability) {
        int tick = barStart + static_cast<int>(i) * sixteenth;
        tick = applySwing(tick, TICKS_PER_BEAT, profile.swingAmount);

        int timingOffset = static_cast<int>(
            (dist(rng) - 0.5f) * profile.timingHumanize * TICKS_PER_BEAT);
        int velocity = velDist(rng);
        if (i == 0)
          velocity = std::min(127, velocity + 15);

        allHits.push_back({"kick", GMDrum::KICK, tick, sixteenth, velocity,
                           false, timingOffset});
        instrumentsUsed.insert("kick");
      }
    }

    // Snare pattern
    for (size_t i = 0; i < profile.snarePattern.size(); ++i) {
      if (profile.snarePattern[i] && dist(rng) > profile.dropProbability) {
        int tick = barStart + static_cast<int>(i) * sixteenth;
        tick = applySwing(tick, TICKS_PER_BEAT, profile.swingAmount);

        int timingOffset = static_cast<int>(
            (dist(rng) - 0.5f) * profile.timingHumanize * TICKS_PER_BEAT);
        int velocity = velDist(rng);
        if (i == 4 || i == 12)
          velocity = std::min(127, velocity + 10);

        allHits.push_back({"snare", GMDrum::SNARE, tick, sixteenth, velocity,
                           false, timingOffset});
        instrumentsUsed.insert("snare");
      }
    }

    // Hi-hats
    std::vector<std::tuple<int, std::string, int>> hatHits;
    if (groove == GrooveType::Trap) {
      hatHits = generateTrapHats(ticksPerBar, profile);
    } else {
      hatHits = generateHatPattern(ticksPerBar, profile, groove);
    }

    for (auto &[tick, instr, vel] : hatHits) {
      int actualTick = barStart + tick;
      actualTick = applySwing(actualTick, TICKS_PER_BEAT, profile.swingAmount);
      int timingOffset = static_cast<int>(
          (dist(rng) - 0.5f) * profile.timingHumanize * TICKS_PER_BEAT * 0.5f);

      int midiNote =
          (instr == "open_hat") ? GMDrum::OPEN_HAT : GMDrum::CLOSED_HAT;

      allHits.push_back({instr, midiNote, actualTick, TICKS_PER_BEAT / 8, vel,
                         false, timingOffset});
      instrumentsUsed.insert(instr);
    }

    // Ghost notes
    if (dist(rng) < profile.ghostNoteProbability) {
      std::vector<int> ghostPositions = {2, 6, 10, 14};
      int ghostPos = ghostPositions[rng() % ghostPositions.size()];
      int tick = barStart + ghostPos * sixteenth;

      allHits.push_back({"snare", GMDrum::SNARE, tick, sixteenth,
                         static_cast<int>(25 + dist(rng) * 20), true, 0});
    }
  }

  // Sort by time
  std::sort(
      allHits.begin(), allHits.end(), [](const DrumHit &a, const DrumHit &b) {
        return (a.startTick + a.timingOffset) < (b.startTick + b.timingOffset);
      });

  RhythmOutput output;
  output.hits = allHits;
  output.config = config;
  output.grooveUsed = groove;
  output.densityUsed = density;
  output.totalTicks = totalTicks;
  output.instrumentsUsed = instrumentsUsed;

  return output;
}

RhythmOutput RhythmEngine::generateIntro(const std::string &emotion, int bars,
                                         int tempo, const std::string &genre) {
  RhythmConfig config;
  config.emotion = emotion;
  config.bars = bars;
  config.tempoBpm = tempo;
  config.genre = genre;
  config.densityOverride = PatternDensity::Minimal;
  config.includeFills = false;
  return generate(config);
}

RhythmOutput RhythmEngine::generateBuildup(const std::string &emotion, int bars,
                                           int tempo,
                                           const std::string &genre) {
  RhythmConfig config;
  config.emotion = emotion;
  config.bars = bars;
  config.tempoBpm = tempo;
  config.genre = genre;
  config.densityOverride = PatternDensity::Busy;
  config.includeFills = true;
  config.fillOnBar = {bars - 1};
  return generate(config);
}

} // namespace kelly
